components/eam/src/physics/cam/modal_aer_opt.F90

```diff
@@ -585,17 +613,32 @@ subroutine modal_aero_sw(list_idx, state, pbuf, nnite, idxnite, is_cmip6_volc, e
    call rad_cnst_get_info(list_idx, nmodes=nmodes)
 
    if (list_idx == 0) then
-      ! water uptake and wet radius for the climate list has already been calculated
-      call pbuf_get_field(pbuf, dgnumwet_idx, dgnumwet_m)
-      call pbuf_get_field(pbuf, qaerwat_idx,  qaerwat_m)
+
+      ! Calculate aerosol size distribution parameters and aerosol water uptake
+      ! For prescribed aerosol codes
+      if (clim_modal_aero .and. .not. prog_modal_aero) then
+         ! diagnostic aerosol size calculations
+         call modal_aero_calcsize_diag(state, pbuf, list_idx, dgnumdry_m)
```

Comment:
@singhbalwinder there might be another potential issue, if I've traced the code correctly.

1. Here, dgnumdry_m is passed to modal_aero_calcsize_diag.
2. dgnumdry_m was allocated during initialization as a module-level array that persists throughout the run.
3. But modal_aero_calcsize_diag calls `allocate(dgnum_m(pcols,pver,nmodes), stat=stat)` for the given set of arguments to modal_aero_calcsize_diag that we see at this line.
4. Thus, it appears that dgnumdry_m is allocated at each call to modal_aero_calcsize_diag. But because this module assumes dgnumdry_m is allocated just once at initialization, these subsequent allocations are not deallocated. And since dgnum_m has type `pointer` rather than `allocatable`, it won't be auto-deallocated.

Sorry if I got that wrong.

Reply:
Thanks Andrew. This indeed seems like a bug. My goal was to remove “calcsize_diag” entirely and replace it with “calcsize_sub” but there isn’t enough time now to remove it and test it throughly. I was testing “calcsize_sub” routine all the time because it is one relevant for prognostic aerosols. “calcsize_diag” is never called for prognostic aerosols now.

This routine is called from different places in the code with different configurations. My intention here was to get dry size (dgnum) using the "fresh state” even with list_dix = 0. The extra allocate should be removed and we should expect dgnum to be already allocated in this routine. Thanks again for catching this!